fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/protocol/ProtocolChain.java

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 Oath Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.web.apirequest.generator.metric.protocol;
+
+import com.yahoo.bard.webservice.data.metric.LogicalMetric;
+import com.yahoo.bard.webservice.data.metric.LogicalMetricInfo;
+import com.yahoo.bard.webservice.data.metric.protocol.Protocol;
+import com.yahoo.bard.webservice.data.metric.protocol.ProtocolMetric;
+import com.yahoo.bard.webservice.data.metric.protocol.UnknownProtocolValueException;
+import com.yahoo.bard.webservice.web.apirequest.generator.metric.antlr.ProtocolAntlrApiMetricParser;
+import com.yahoo.bard.webservice.web.apirequest.metrics.ApiMetric;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+/**
+ * Protocol Chain attempts to apply a set of protocols to a logical metric to produce additional metrics.
+ *
+ */
+public class ProtocolChain {
+    private static final Logger LOG = LoggerFactory.getLogger(ProtocolAntlrApiMetricParser.class);
+
+    private final boolean strictValidation;
+
+    List<Protocol> protocols;
+
+    /**
+     * Constructor.
+     *
+     * @param protocols  The chain of protocol instances to be attempted on ApiMetrics.
+     * @param strictValidation  throw an error if a protocol's triggering parameters is present but the protocol
+     * cannot be accepted.
+     */
+    public ProtocolChain(List<Protocol> protocols, boolean strictValidation) {
+        this.protocols = protocols;
+        this.strictValidation = strictValidation;
+    }
+
+    /**
+     * Constructor.
+     * Strict validation defaulted to off;
+     *
+     * @param protocols  The chain of protocol instances to be attempted on ApiMetrics.
+     */
+    public ProtocolChain(List<Protocol> protocols) {
+        this(protocols, false);
+    }
+
+    /**
+     * Dummy implementation.
+     *
+     * @param resultMetadata  The metadata for the expected metric
+     * @param apiMetric  The metric described in the apiRequest
+     * @param fromMetric  The base LogicalMetric to be transformed
+     *
+     * @return  The logical metric after all Protocols have been applied.
+     * @throws UnknownProtocolValueException if the protocol data from the query cannot be applied
+     */
+    public LogicalMetric applyProtocols(
+            LogicalMetricInfo resultMetadata,
+            ApiMetric apiMetric,
+            LogicalMetric fromMetric
+    )
+            throws UnknownProtocolValueException {
+        LogicalMetric soFar = fromMetric;
+        for (Protocol p: protocols) {
```

Comment:
I think if a core parameter is passed that does not trigger on any core parameters we will silently ignore it. Since we use the raw name it will never be apparent that the core parameter is ignored. I think we need to have some sort of validation against this.

Reply:
There is no useful way to validate it without creating an entire metadata piece for every possible parameter on every possible lookup and even then there may be parameters sent which are politely ignored for validation reasons.  Basically, this 'indifference to surplus signal' is intentional and I think architecturally not worth the cost of avoiding.

Reply:
ApiMetric has a map of protocol core parameter to value. We can ensure all values in that map correctly match a protocol in the system or else we reject the query. I definitely don't think silently ignoring a typo on a core parameter which causes a totally different calculation that the use intended is correct.

Reply:
Issue on the subject filed here: https://github.com/yahoo/fili/issues/1025